transport/src/main/java/io/netty/bootstrap/ChannelInitializerExtensions.java

```diff
@@ -0,0 +1,116 @@
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.bootstrap;
+
+import io.netty.util.internal.SystemPropertyUtil;
+import io.netty.util.internal.logging.InternalLogLevel;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.ServiceLoader;
+
+/**
+ * The configurable facade that decides what {@link ChannelInitializerExtension}s to load and where to find them.
+ */
+abstract class ChannelInitializerExtensions {
+    private static final InternalLogger logger = InternalLoggerFactory.getInstance(ChannelInitializerExtensions.class);
+    private static volatile ChannelInitializerExtensions implementation;
+
+    private ChannelInitializerExtensions() {
+    }
+
+    /**
+     * Get the configuration extensions, which is a no-op implementation by default,
+     * or a service-loading implementation if the {@code io.netty.bootstrap.extensions} system property is
+     * {@code serviceload}.
+     */
+    static ChannelInitializerExtensions getExtensions() {
+        ChannelInitializerExtensions impl = implementation;
+        if (impl == null) {
+            synchronized (ChannelInitializerExtensions.class) {
+                impl = implementation;
+                if (impl != null) {
+                    return impl;
+                }
+                String extensionProp = SystemPropertyUtil.get(ChannelInitializerExtension.EXTENSIONS_SYSTEM_PROPERTY);
+                logger.debug("-Dio.netty.bootstrap.extensions: {}", extensionProp);
+                final Collection<ChannelInitializerExtension> extensions;
+                if ("serviceload".equalsIgnoreCase(extensionProp)) {
+                    extensions = serviceLoadExtensions(InternalLogLevel.DEBUG);
+                } else if ("log".equalsIgnoreCase(extensionProp)) {
+                    serviceLoadExtensions(InternalLogLevel.INFO);
+                    extensions = Collections.emptyList();
+                } else {
+                    extensions = Collections.emptyList();
+                }
+                implementation = impl = new LoadedExtensions(extensions);
+            }
+        }
+        return impl;
+    }
+
+    private static Collection<ChannelInitializerExtension> serviceLoadExtensions(InternalLogLevel logLevel) {
+        ServiceLoader<ChannelInitializerExtension> loader = ServiceLoader.load(ChannelInitializerExtension.class);
```

Comment:
`ServiceLoader.load(Class)` uses the context class loader, which is dependent on the current thread. It isn't really appropriate to save the result since it is, by definition, context-dependent.

We used `ServiceLoader.load(Class, ClassLoader)` [in gRPC](https://github.com/grpc/grpc-java/blob/6335e0be3ef2b0195eae874a9e1a65a031328646/api/src/main/java/io/grpc/ServiceProviders.java#L107), which provides consistent results. However, things like http://aries.apache.org/modules/spi-fly.html didn't support passing the ClassLoader, so if the ClassLoader call returns nothing, we try the context-based approach. If SPI Fly is in use, then it will return consistent results...

Reply:
Where do you get the class loader parameter from? Do you pick a default somewhere, or must it always be explicitly passed in?

Reply:
We pick it from the class doing the loading. So `getClass().getClassLoader()` in this case.